doc/world_format.md

```diff
@@ -0,0 +1,658 @@
+
+Before map format version 29:
+* `u32` `timestamp`
+    * Same meaning as the timestamp further up
+
+* `u8` `name-id-mapping` version
+    * Always 0
+
+* `u16` `num_name_id_mappings`
+    * foreach `num_name_id_mappings`:
+        * `u16` `id`
+        * `u16` `name_len`
+        * `u8[name_len]` `name`
+
+EOF.
+
+# Format of Nodes
+
+A node is composed of the `u8` fields `param0`, `param1` and `param2`.
+
+Before map format version 24:
+* The content id of a node is determined as so:
+    * If `param0` < `0x80`, `content_id = param0`
+    * Otherwise, `content_id = (param0<<4) + (param2>>4)`
+
+Since map format version 24:
+* The content id of a node is `param0`.
+
+The purpose of `param1` and `param2` depend on the definition of the node.
+
+# Name-ID-Mapping
+
+The mapping maps node content ids to node names.
+
+# Node Metadata Format (Before Map Format Version 23)
+
+The node metadata is serialized depending on the `type_id` field.
+
+`1`: Generic metadata
+* serialized inventory
+* `u32` `len`
+* `u8[len]` `text`
+* `u16` `len`
+* `u8[len]` `owner`
+* `u16` `len`
+* `u8[len]` `infotext`
+* `u16` `len`
+* `u8[len]` inventory drawspec
+* `u8` `allow_text_input` (bool)
+* `u8` `removal_disabled` (bool)
+* `u8` `enforce_owner` (bool)
+* `u32` `num_vars`
+* foreach `num_vars`:
+    * `u16` `len`
+    * `u8[len]` `name`
+    * `u32` `len`
+    * `u8[len]` `value`
+
+`14`: Sign metadata
+* `u16` `text_len`
+* `u8[text_len]` `text`
+
+`15`: Chest metadata
+* serialized inventory
+
+`16`: Furnace metadata
+* To be determined
+
+`17`: Locked Chest metadata
+* `u16` `len`
+* `u8[len]` `owner`
+* serialized inventory
+
+# Static Objects
+
+Static objects are persistent freely moving objects in the world.
+
+Object types:
+`1`: Test object
+`7`: LuaEntity:
+* `u8` `compatibility_byte` (always 1)
+* `u16` `len`
+* `u8[len]` entity name
+* `u32` `len`
+* `u8[len]` static data
+* `s16` `hp`
+* `s32` velocity.x * 10000
+* `s32` velocity.y * 10000
+* `s32` velocity.z * 10000
+* `s32` yaw * 1000
+
+Since protocol version 37:
+* `u8` `version2` (=1)
+* `s32` pitch * 1000
+* `s32` roll * 1000
+
+# Itemstring Format
+
+Examples:
+* `'default:dirt 5'`
+* `'default:pick_wood 21323'`
+* `'"default:apple" 2'`
+* `'default:apple'`
+
+Older formats:
+* `'node "default:dirt" 5'`
+* `'NodeItem default:dirt 5'`
+* `'ToolItem WPick 21323'`
+
+The wear value in tools is 0...65535
+
+# Inventory Serialization Format
+
+* The inventory serialization format is line-based
+* The newline character used is `\n`
+* The end condition of a serialized inventory is always `EndInventory\n`
+* All the slots in a list must always be serialized.
+
+## Example
+
+    List foo 4
+    Item default:sapling
+    Item default:sword_stone 1 10647
+    Item default:dirt 99
+    Empty
+    EndInventoryList
+    List bar 9
+    Empty
+    Empty
+    Empty
+    Empty
+    Empty
+    Empty
+    Empty
+    Empty
+    Empty
+    EndInventoryList
+    EndInventory
+
+# Table of Contents
```

Comment:
I feel like such a table of contents is (1) redundant and (2) probably will eventually be out-of-sync with the entire document. I think generating a ToC is the job of Markdown renderers.

Reply:
GitHub shows a table of contents in the top left

Reply:
Ah. I didn't know that, I'll remove them